tests/Unit/Encoders/JsonTest.php

```diff
@@ -46,7 +47,15 @@ public function testEncodeIgnoreSpanWhenEncodingFails()
         // this will generate a malformed UTF-8 string
         $span->setTag('invalid', hex2bin('37f2bef0ab085308'));
 
+        $logger = $this->prophesize(LoggerInterface::class);
+        $logger
+            ->debug(
+                'Failed to json-encode span: Malformed UTF-8 characters, possibly incorrectly encoded'
+            )
+            ->shouldBeCalled();
+
         $jsonEncoder = new Json();
+        $jsonEncoder->setLogger($logger->reveal());
```

Comment:
Could the logger be set in constructor one line above?

Reply:
I agree it would be better to allow injection via constructor.

How about

```diff
diff --git a/src/DDTrace/Encoders/Json.php b/src/DDTrace/Encoders/Json.php
index 01876e6..8b8c8f2 100644
--- a/src/DDTrace/Encoders/Json.php
+++ b/src/DDTrace/Encoders/Json.php
@@ -4,12 +4,19 @@ namespace DDTrace\Encoders;

 use DDTrace\Encoder;
 use DDTrace\Span;
-use Psr\Log\LoggerAwareInterface;
-use Psr\Log\LoggerAwareTrait;
+use Psr\Log;

 final class Json implements Encoder
 {
-    use LoggerAwareTrait;
+    /**
+     * @var Log\LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(Log\LoggerInterface $logger = null)
+    {
+        $this->logger = $logger ?: new Log\NullLogger();
+    }

     /**
      * {@inheritdoc}
diff --git a/tests/Unit/Encoders/JsonTest.php b/tests/Unit/Encoders/JsonTest.php
index 5db16ef..1f46924 100644
--- a/tests/Unit/Encoders/JsonTest.php
+++ b/tests/Unit/Encoders/JsonTest.php
@@ -54,8 +54,7 @@ JSON;
             )
             ->shouldBeCalled();

-        $jsonEncoder = new Json();
-        $jsonEncoder->setLogger($logger->reveal());
+        $jsonEncoder = new Json($logger->reveal());
         $encodedTrace = $jsonEncoder->encodeTraces([[$span, $span]]);
         $this->assertEquals($expectedPayload, $encodedTrace);
     }
```

?